Add SyncCacheApi implementation over PlayRedis class.

felipebonezi opened the pull request:
Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

Issue Ref: #6

felipebonezi committed:
Add SyncCacheApi implementation over PlayRedis class.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Fix travis checkstyle problems with final variables.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Fix travis checkstyle problems with final variables.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Refactor sync implementation.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Fix unit tests.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Implement unit test to increase CodeCov.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Implement unit test to increase CodeCov.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi commented:
@thibaultmeyer Can I implement the `AsyncCacheAPI` in this same Pull Request?
Or is it best to create a new one?

thibaultmeyer commented:
Hello,

You can implements it in the same merge request. 

Thank you

felipebonezi committed:
Refactor async implementation.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Fix code checkstyle with final parameters.

Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi commented:
@thibaultmeyer I've done the implementation, but I don't know why the CodeCov isn't working!
Can you check?

felipebonezi commented:
> @thibaultmeyer I've done the implementation, but I don't know why the CodeCov isn't working!
> Can you check?

All checks have passed @thibaultmeyer 🎉

thibaultmeyer commented on src/main/java/com/zero_x_baadf00d/play/module/redis/cache/AsyncCacheRedisImpl.java:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @since 20.11.05
+     */
+    private final Set<String> keys = new HashSet<>(100);
```
I'm not sure. This mechanism will not work if several instances of the application are running. The simplest would be, in my opinion, to use the `getConnection` method to execute the `KEYS` method which allows to find all the keys matching needed patter.

You can use a hardcoded prefix or allow user to change it in the configuration file. By example if you use the prefix `cache.`, the method `removeAll` will use Redis connection to find all keys matching `cache.*` and delete them.

Redis doc : https://redis.io/commands/keys


**Example**
```java
@Override
public CompletionStage<Done> removeAll() {
   return supplyAsync(() -> {
        this.playRedis.getConnection(0).keys("search pattern").forEach(this::remove);
        return Done.done();
    });
}
```

thibaultmeyer commented on src/main/java/com/zero_x_baadf00d/play/module/redis/cache/AsyncCacheRedisImpl.java:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @since 20.11.05
+     */
+    private final Set<String> keys = new HashSet<>(100);
```
I'm not sure. This mechanism will not work if several instances of the application are running. The simplest would be, in my opinion, to use the `getConnection` method to execute the `KEYS` method which allows to find all the keys matching needed patter.

You can use a hardcoded prefix or allow user to change it in the configuration file. By example if you use the prefix `cache.`, the method `removeAll` will use Redis connection to find all keys matching `cache.*` and delete them.

Redis doc : https://redis.io/commands/keys


**Example**
```java
@Override
public CompletionStage<Done> removeAll() {
   return supplyAsync(() -> {
        this.playRedis.getConnection(0).keys("search pattern").forEach(this::remove);
        return Done.done();
    });
}
```

felipebonezi replied:
@thibaultmeyer thanks, I think it makes sense.

felipebonezi committed:
Improve removeAll method to use Redis native method.

We're using keys() method to search all keys saved in Redis with a prefix pattern.

https://redis.io/commands/keys
Signed-off-by: Felipe Bonezi <1410319+felipebonezi@users.noreply.github.com>

felipebonezi committed:
Fix README.

thibaultmeyer commented on src/main/java/com/zero_x_baadf00d/play/module/redis/cache/AsyncCacheRedisImpl.java:
```diff
@@ -0,0 +1,161 @@
+            try (final Jedis connection = this.asPlayRedis().getConnection(0)) {
+                // Search all keys in Redis.
+                // https://redis.io/commands/keys
+                connection.keys(DEFAULT_PREFIX + "??").forEach(this::remove);
```
I think you have to replace `??` with `*`. According to the documentation `??` will only match a two characters word.

By example, if you have the following keys :
- play.cache.p1
- play.cache.cn
- play.cache.index
- play.cache.dashboard

The pattern `DEFAULT_PREFIX + "??"` will only match "play.cache.p1" and "play.cache.cn". The pattern  `DEFAULT_PREFIX + "*"` will match all keys beginning with DEFAULT_PREFIX

felipebonezi replied:
You're right!

felipebonezi committed:
Fix pattern search in Redis.

thibaultmeyer reviewed (approved):
Nice work. Thank you !
